File: linked_list/linkedCycleII.py
from typing import Optional

class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None


def detectCycle( head: Optional[ListNode]) -> Optional[ListNode]:

        fast = head
        slow = head

        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

            if slow == fast:
                break
        
        length = 0

        while slow.next != fast:
            slow = slow.next
            length += 1

        print(length)

        fast = head

        while length != 0:
            fast = fast.next
            length -= 1  # decrement the length in each iteration

        return fast

a = ListNode(3)
b = ListNode(2)

c = ListNode(0)

d = ListNode(-4)

a.next = b
b.next = c
c.next = d
d.next = a

print(detectCycle(a).val)

File: linked_list/linkedCycleII.java
class linkedCycleII {
    static class ListNode {
        int val;
        ListNode next; 

        private ListNode (int val) {
            this.val = val;
            this.next = null;
        }
    }

    public static void main(String[] args) {
        ListNode a = new ListNode(3);
        ListNode b = new ListNode(2);
        ListNode c = new ListNode(0);
        ListNode d = new ListNode(-4);

        a.next = b;
        b.next = c;
        c.next = d;
        d.next = b;

        ListNode one = new ListNode(1);
        ListNode two = new ListNode(2);

        one.next = two;
        two.next = one;



        ListNode ans = detectCycle(a);

        System.out.println(ans.val);
    }

    public static ListNode detectCycle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast.next != null && fast != null) {
            fast = fast.next.next;
            slow = slow.next;

            if (slow == fast){
                break;
            }
        }

        int lengthCycle = 0;

        do {
            slow = slow.next;
            lengthCycle++;
        } while (slow != fast);

        System.out.println(lengthCycle);

        //ListNode curr = head;

        while (lengthCycle != 0) {
            slow = slow.next;
            lengthCycle--;
        }

    
        return slow;
    } 

    
}
